File: python/problems/leetcode/0020_valid_parentheses/solution.py
"""LeetCode 20: Valid Parentheses."""

from __future__ import annotations


class Solution:
    def isValid(self, s: str) -> bool:
        st: list[str] = []
        pair = {")": "(", "]": "[", "}": "{"}
        for ch in s:
            if ch in "([{":
                st.append(ch)
            else:
                if not st or st[-1] != pair[ch]:
                    return False
                st.pop()
        return len(st) == 0


if __name__ == "__main__":
    assert Solution().isValid("()[]{}")
    assert not Solution().isValid("(]")
    print("0020 OK")

File: cpp/problems/leetcode/0020_valid_parentheses/solution.cpp
// LeetCode 20
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        vector<char> st;
        unordered_map<char, char> p{{')', '('}, {']', '['}, {'}', '{'}};
        for (char ch : s) {
            if (ch == '(' || ch == '[' || ch == '{')
                st.push_back(ch);
            else {
                if (st.empty() || st.back() != p[ch]) return false;
                st.pop_back();
            }
        }
        return st.empty();
    }
};

int main() {
    assert(Solution().isValid("()[]{}"));
    assert(!Solution().isValid("(]"));
    cout << "0020 OK" << endl;
    return 0;
}
